### PR TITLE
units::volume and units::mass readers for mandatory and optional

### PR DESCRIPTION
**Pull request by KorGgenT**
#### Summary

```SUMMARY: Infrastructure "implement volume_reader and mass_reader for read and write"```
#### Purpose of change

The purpose of this is to make it simple to read and write various units using optional and mandatory from generic_factory, and also able to serialize easily. The usage for this is pretty straightforward: just add one of the readers as a parameter to optional or mandatory. this will avoid seg faults and having to  re implement reading json. The reader does _not_ support legacy units, and we should eventually get away from using legacy 1 = 250 anyway.
#### Describe alternatives you've considered

writing some kind of helper funciton i suppose? that's basically what this is though
#### Testing

Ran the game in vanilla, found some errors. talked to anothersimulacrum about it and he PRed several changes already. This needs to have tests run to make sure that in-repo mods are all prepared for it, as I changed one of the loading functions to use this new reader as a proof of concept.
It would be cool to use this for items and monsters, but those have their own specialized classes rather than using generic_factory.

### CODE REVIEW

**Comment by KorGgenT** [comment]
ran cata tests with every single mod individually and didn't see any errors. i don't know what's up with the template error but i've got some ideas. in the meantime i'm undrafting this as it works perfectly

File: src/units.cpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+void units::mass::serialize( JsonOut &jsout ) const
+{
+    if( value_ % 1000000 == 0 ) {
+        jsout.write( string_format( "%d kg", value_ ) );
+    } else if( value_ % 1000 == 0 ) {
+        jsout.write( string_format( "%d g", value_ ) );
+    } else {
+        jsout.write( string_format( "%d mg", value_ ) );
+    }
+}
```
Comment (kevingranade) [inline src/units.cpp]: See https://stackoverflow.com/questions/25594644/warning-specialization-of-template-in-different-namespace
tl;dr it seems to be a gcc 5.4 bug.
I think you want something like:
```
namespace units
{
template<>
void volume::serialize( JsonOut &jsout ) const
{
    if( value_ % 1000 == 0 ) {
        jsout.write( string_format( "%d L", value_ ) );
    } else {
        jsout.write( string_format( "%d ml", value_ ) );
    }
}

template<>
void mass::serialize( JsonOut &jsout ) const
{
    if( value_ % 1000000 == 0 ) {
        jsout.write( string_format( "%d kg", value_ ) );
    } else if( value_ % 1000 == 0 ) {
        jsout.write( string_format( "%d g", value_ ) );
    } else {
        jsout.write( string_format( "%d mg", value_ ) );
    }
}
};
```
